### PR TITLE
fix(provider): fix SAML passport authentication strategy

### PR DESCRIPTION
**Pull request by RaghavaroraSF**
## Description

Fixed the npm version because it was deprecated

GH-131

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Intermediate change (work in progress)

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist:

- [ ] Performed a self-review of my own code
- [ ] npm test passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [ ] Code conforms with the style guide
- [ ] API Documentation in code was updated
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Comment by yeshamavani** [comment]
Rollback the changes that are not required anymore because of the library change

File: src/strategies/SAML/saml-strategy-factory-provider.ts [hunk 1]
```diff
@@ -52,13 +54,15 @@ export class SamlStrategyFactoryProvider
...
+        options,
+        logoutVerify as VerifyWithRequest,
         // eslint-disable-next-line @typescript-eslint/no-misused-promises
         func,
       );
     } else {
       strategy = new Strategy(
-        options as SamlConfig,
+        options,
+        logoutVerify as unknown as VerifyWithoutRequest,
```
Comment (yeshamavani) [inline src/strategies/SAML/saml-strategy-factory-provider.ts:65]: why is this required

File: src/strategies/SAML/saml-strategy-factory-provider.ts [hunk 2]
```diff
@@ -94,3 +97,10 @@ export class SamlStrategyFactoryProvider
     }
   }
 }
+function logoutVerify(
+  req: Request<AnyObject, AnyObject, Record<string, AnyObject>>,
+  profile: Profile | null,
+  done: VerifiedCallback,
+): void {
+  throw new Error('Function not implemented.');
+}
```
Comment (yeshamavani) [inline src/strategies/SAML/saml-strategy-factory-provider.ts:106]: What is the use of this function 
mention it in comments 
and how this error can be handled provide an example if possible

**Review by yeshamavani — COMMENTED** [review]
Also add the docs to readme

File: README.md [hunk 3]
```diff
@@ -2595,6 +2596,358 @@ this.bind(VerifyBindings.BEARER_SIGNUP_VERIFY_PROVIDER).toProvider(
...
+      clientID: process.env.SAML_CLIENT_ID,
+      clientSecret: process.env.SAML_CLIENT_SECRET,
+      tokenURL: process.env.SAML_TOKEN_URL,
+    },
+    queryGen('query'),
+  )
+  @authorize({permissions: ['*']})
+  @oas.deprecated()
+  @get('/saml', {
+    description: 'POST Call for saml based login',
```
Comment (yeshamavani) [inline README.md:2805]: POST call ?

File: README.md [hunk 4]
```diff
@@ -2595,6 +2596,358 @@ this.bind(VerifyBindings.BEARER_SIGNUP_VERIFY_PROVIDER).toProvider(
...
+      const token = await this.getAuthCode(client, user);
+      response.redirect(`${client.redirectUrl}?code=${token}`);
+    } catch (error) {
+      this.logger.error(error);
+      throw new HttpErrors.Unauthorized(AuthErrorKeys.InvalidCredentials);
+    }
+  }
+```
+
+Please note above that we are creating two new APIs for SAML. The first one is for UI clients to hit. We are authenticating client as well, then passing the details to the SAML. Then, the actual authentication is done by SAML authorization url, which redirects to the second API we created after success. The first API method body is empty as we do not need to handle its response. The SAML provider in this package will do the redirection for you automatically.
```
Comment (yeshamavani) [inline README.md:2923]: why three API

File: README.md [hunk 5]
```diff
@@ -2595,6 +2596,358 @@ this.bind(VerifyBindings.BEARER_SIGNUP_VERIFY_PROVIDER).toProvider(
...
+  private readonly getCurrentUser: Getter<User>,
+```
+
+The `logoutVerify` function is used in the node-saml library as a part of the Passport SAML authentication process. This function is used to verify the authenticity of a SAML logout request.
+The logout process in SAML is used to end the user's session on the service provider, and the logoutVerify function is used to verify that the logout request is coming from a trusted IdP.
+The implementation of the logoutVerify function may vary depending on the specific requirements and the security constraints of the application. It is typically used to verify the signature on the logout request, to ensure that the request has not been tampered with, and to extract the user's identity information from the request.
+
+```ts
+function logoutVerify(
+  req: Request<AnyObject, AnyObject, Record<string, AnyObject>>,
```
Comment (yeshamavani) [inline README.md:2938]: Can we create any specific types instead of AnyObjects?

**Review by yeshamavani — CHANGES_REQUESTED** [review]
Rebase the PR 
Also this will go as a feature and not fix

File: README.md (hunk 3 above)
Comment (RaghavaroraSF) [reply README.md:2805]: resolved

File: README.md (hunk 4 above)
Comment (RaghavaroraSF) [reply README.md:2923]: resolved

File: README.md (hunk 5 above)
Comment (RaghavaroraSF) [reply README.md:2938]: Speciifc type was giving error so as to resolve the error used this.
